(WIP) V3

ravanscafi opened the pull request:
## To-do
- [X] PHP 7.1 
- [X] Code Standards
- [X] Fix directories
- [x] Drop connection `pool` (see https://github.com/mongodb/mongo-php-driver/issues/688#issuecomment-349193875)
- [x] Better `update` method (including #122)
- [x] Allow modification of array values directly
- [x] Stop saving fields with value `null`
- [x] Replace `attributes` by other field
- [x] Avoid `first` queries with `null`
- [X] Better relations support (#136)
- [X] Relationship cache on model
- [x] `created_at` and `updated_at` management on embedded relations
- [X] Add to [PHP libraries](https://docs.mongodb.com/ecosystem/drivers/php-libraries/) (see https://github.com/mongodb/docs-ecosystem/pull/463)
- [x] Suggest dependent packages
- [x] Review `Cursor` and `CacheableCursor` (#115)
    - [x] Drop CacheableCursor
- [x] Review `DataMapper` as it does not follow Data Mapper pattern correctly (perhaps drop it?)
- [X] Leverage [`Serializable`](https://secure.php.net/manual/en/class.mongodb-bson-serializable.php)/[`Unserializable`](https://secure.php.net/manual/en/class.mongodb-bson-unserializable.php) classes on saving/retrieving objects implementing [Persistable](http://php.net/manual/pt_BR/class.mongodb-bson-persistable.php).
- [x] Create `ModelInterface`
- [x] Update Readme
- [x] New documentation

## Dependent Packages
- [ ] [Mongolid Laravel](https://github.com/leroy-merlin-br/mongolid-laravel) (see https://github.com/leroy-merlin-br/mongolid-laravel/pull/126)

## Maybe (not breaking changes, so that can be implemented later)
- [ ] Hidden attributes when converting to json #25
- [ ] `refresh()` (#40)
- [ ] `complete()`/`incomplete()` and saving exceptions (for models retrieved with projections)
- [ ] Better relationship support (some reverse relationships that are missing, using laravel collections,  methods like `has`, `is`, etc.)

carusogabriel committed:
Use LMBR Coding Standard

Commit:
Move files and drop bootstrap file

Commit:
Ignore some files when exporting package

Commit:
Merge branch 'master' into coding-standard

Commit:
Require PHP >= 7.1

ravanscafi committed:
Merge pull request #129 from leroy-merlin-br/coding-standard

Use LMBR Coding Standard

Commit:
Merge branch 'v3' into improve-exports

Commit:
Fix CS for new structure

Commit:
Update readme, license and composer.json files

Commit:
Fix Schema tests namespace

Commit:
Use new mockery syntax

ravanscafi committed:
Commit lock file to speed up installation

ravanscafi committed:
Move unit tests

ravanscafi committed:
Create integration tests folder and testsuite

ravanscafi committed:
Fix CS rules

ravanscafi committed:
Add pre-commit hook

ravanscafi committed:
Create integration test to ensure no regressions later

ravanscafi committed:
Fix travis

ravanscafi committed:
Update .gitattributes

ravanscafi committed:
Drop connection pool

The PHP driver uses libmongoc's single-threaded client mode, which does not implement connection pools. There is also no way to implement your own connection pooling using this driver.

Each single-threaded libmongoc client object maintains at most one socket connection per mongod or mongos server. The relationship between MongoDB\Driver\Manager objects and libmongoc client objects is discussed in Connection Handling in the driver documentation.

Connection pools should not be needed for any single-threaded application, unless you're using an async PHP framework (e.g. Amp, ReactPHP). In those cases, you'd probably need to implement your own driver as we provide no way to integrate libmongoc's IO with a PHP event loop. Multi-threaded PHP applications (e.g. pthreads, forking) are not supported by the driver.
https://github.com/mongodb/mongo-php-driver/issues/688#issuecomment-349193875

Commit:
Use atomic updates for update() method of data mapper

Commit:
Allow modification of array values directly

It is now possible to use `$model->array['key'] = 'value';`

Commit:
Allow integration testes to be accounted on coverage results

ravanscafi committed:
Merge pull request #134 from leroy-merlin-br/update-method

Use atomic updates for update() method of data mapper

Commit:
Move Manager to Connections Namespace

Commit:
Move ActiveRecord to Model Namespace

Commit:
Prevent query 'first(null)' to hit the database

This simple change improve relationship checks, where for example a `referencesOne` with
no value would hit the database previously.

Commit:
Make mutable field protected

Commit:
Small improvements

Commit:
Rename model attributes and original attributes

Commit:
Create integration tests for relationships by reference

Commit:
Improve coverage a bit

ravanscafi committed:
Update composer and coveralls dependency

Commit:
Suggest installing mongolid-laravel package

Commit:
Update travis CI script and allow PHP 7.3 to run

Commit:
Remove get_class() and improve CursorInterface

Commit:
Replace Ioc:instance() calls

Commit:
Simplify protected helpers

Commit:
Replace self::class with static::class and makeWith with make

Commit:
Revert change that is breaking on CI

Commit:
Change document attributes based on team feedback

Commit:
Fix mutator names for snake_case attributes

Commit:
Rename traits, interfaces and abstract classes to follow code standard

Commit:
Improve a bit the relation between AbstractActiveRecord and HasAttributesTrait

Commit:
New draft for relations

Commit:
Fix HasRelationsTraitTest tests

Commit:
Improve how embedded relations are returned

Commit:
Improve tests scenarios

Commit:
Drop unused method

Commit:
Improve polymorph docs a bit

Commit:
Allow relations by other reference field

Commit:
Infer relation names

Commit:
Create RelationInterface and validate attribute access of methods

Commit:
Flip attributes to follow the same order as parent

Commit:
Rename exception to be clearer

Commit:
Prevent field name from being the same as the relation name

Commit:
Fixes based on review

Commit:
Add a test for old reserved words

Commit:
Improve attributes test

Commit:
New cache strategy for relations

Commit:
Invalidate relation cache when unsetting field

Commit:
Invalidate relation cache when setting field directly

Commit:
Simplify logic relations' cached results

Commit:
Prevent EmbedsOne and ReferencesOne to have more than one relation

ravanscafi committed:
Merge pull request #136 from leroy-merlin-br/new-relations

New relations

Commit:
Fix regression on DataMapper Events

Commit:
Fix regression where DocumentEmbedder would not trigger setDocumentAttribute

Commit:
Only create relations with objects, embed only attributes and create new helper methods

Commit:
Parse document to consider fields when embedding models

Commit:
Add new test to make sure that date queries works

Commit:
Fix LogicException bug (probably mongodb bug?) and add test to it

Commit:
Use CachingIterator to wrap cursor

Commit:
Some improvements on Cursors

Commit:
Remove cacheable cursor implementation

Commit:
Drop cursor factory

Commit:
"Drop" DataMapper pattern

We don't actually implement the pattern neither use it, so there is no need to have it.

Commit:
Implement Persistable interface and simplify logic for parsing documents

Commit:
Merge branch 'master' into v3

Commit:
Improve docker configuration

Commit:
Set schema for broken tests

Commit:
Simplify schema use, drop model assembler and polymorphable interface

There is no more the need to assemble models, or to use polymorph.
Persistable interface takes care of that, with the same funcionality and less performance hit.

Commit:
Drop schemas in favor of bson serialization

Commit:
Update readme

Commit:
Improve test steps

Commit:
Only report coverage for version 7.2 and ensure it runs on 7.3

Commit:
Fix wrong renaming

Commit:
Make fill method static and change signature to allow object to be passed

Commit:
Move stubs to common tests namespace

Commit:
Use stubs for Relations test

Commit:
Create new polymorphable model interface

Commit:
Add __pclass field to projection

ravanscafi committed:
Merge pull request #139 from leroy-merlin-br/new-polymorph

New polymorph

Commit:
Move document embedder to relations namespace

Commit:
Remove DocumentEmbedder and simplify relations with only one model

Commit:
Add Arrayable interface to Cursor interface

Commit:
Simplify query builder test

Commit:
Add test to ensure that cursor preferences are not overriding each other

Commit:
Fix coverage reporting

Commit:
Renaming rawCollection to client

Commit:
Add integration test for BulkWrite

Commit:
Use bulkwrite method from collection to simplify things a bit

Commit:
Drop getManager from Connection as it is not used anymore

Commit:
Create getCollection on Model to simplify interacting with MongoDB Collection directly

Commit:
Rename Ioc to Container

Commit:
Fix method isEmpty of BulkWrite and ensure it is empty after running

Commit:
100% coverage

Commit:
Use implemented getCollection for test stubs

Commit:
Update travis script

Commit:
Change requirements order on readme

Commit:
Use container to create instances on fill method

Commit:
Consider model attributes when checking for polymorph class

Commit:
Unset field when value is null

Commit:
Clean null attributes when retrieving them

Commit:
Fix embedded models being retrieved as arrays instead of objects

As a drawback, it is not possible to retrieve stdClass' fields from database.
They are casted to array (to keep array behavior consistent).

Commit:
Fix test that was breaking on CI

Commit:
Remove unused variable

Commit:
Small tweaks

carusogabriel commented on composer.json:
```diff
@@ -11,31 +11,45 @@
+        "leroy-merlin-br/coding-standard": "^0.1",
+        "mockery/mockery": "^1.2",
+        "php-coveralls/php-coveralls": "^2.1",
+        "phpunit/phpunit": "^6.5",
```
We can use PHPUnit 7 here, as we are now minimum requiring PHP 7.1.

Commit:
Use container to manage timestamps and id on ModelMapper

This improves testing timestamps and id generation on applications using mongolid.

carusogabriel commented:
@ravanscafi What about implementing a `last()` method? The opposite for `first()`.

ravanscafi commented on src/Model/HasAttributesTrait.php:
```diff
@@ -0,0 +1,238 @@
+    public function syncOriginalDocumentAttributes()
+    {
+        try {
+            $this->originalAttributes = unserialize(serialize($this->getDocumentAttributes()));
```
note to self: this takes way too many memory, think about alternatives or drop it altogether.

Commit:
Add a test to ensure that embedded models are correctly updated

carusogabriel commented on README.md:
```diff
@@ -1,85 +1,45 @@
-
-- PHP**7**
+## Requirements
+- PHP **7.1** or superior
```
```suggestion
- PHP **7.1** or higher
```

carusogabriel commented on composer.lock:
```diff
@@ -0,0 +1,3401 @@
+{
```
I don't see the value of adding the `composer.lock` file to a package, as Composer ignores it by default.

ravanscafi committed:
Update phpunit and illuminate dependencies

ravanscafi committed:
Bump php minimum version to 7.2

ravanscafi committed:
Attempt to bump mongodb driver version

ravanscafi committed:
Merge branch 'master' into v3

ravanscafi committed:
Configure mongodb extension on travis

Edukazan committed:
Upgrade illuminate dependencies (#147)

* Upgrade illuminate dependencies

This upgrade is necessary to make mongolid compatible with laravel 7

* Revert "Upgrade illuminate dependencies"

This reverts commit ea0df635cfdd22097af6e014af6fcda034d27c7b.

* Add Illuminate ^7.26 dependencies

Add Illuminate ^7.26 dependencies

* Revert "Add Illuminate ^7.26 dependencies"

This reverts commit e2ffe3a6c8091c8af53f5a66caaf667966670e82.

* Add Illuminate 7 dependencies

Commit:
Update our CS version

Commit:
Allow Laravel 8.*

williamdes committed:
Replace Sami by Doctum

vidalvasconcelos committed:
Merge pull request #150 from wdesportes/master

Replace Sami by Doctum

vidalvasconcelos committed:
feat: remove typos from README.md

vidalvasconcelos committed:
fix: covenant code of conduct typos

vidalvasconcelos committed:
fix: use a heading instead of use emphasis

vidalvasconcelos committed:
Merge pull request #155 from leroy-merlin-br/docs/remove-typos

docs: remove typos from README
